cluster.c

```diff
@@ -889,6 +951,60 @@ RRStatus ShardGroupParse(RedisModuleCtx *ctx, RedisModuleString **argv, int argc
     return RR_ERROR;
 }
 
+/* parses all the shardgroups in a replace call and validates them */
+RRStatus ShardGroupsParse(RedisModuleCtx *ctx, RedisModuleString **argv, int argc, RaftReq *req)
+{
+    if (RedisModule_StringToLongLong(argv[0], &req->r.shardgroups_replace.len) != REDISMODULE_OK) {
+        RedisModule_ReplyWithError(ctx, "ERR invalid shard group message(1)");
+        return RR_ERROR;
+    }
+    req->r.shardgroups_replace.shardgroups =
+            RedisModule_Calloc(req->r.shardgroups_replace.len, sizeof(ShardGroup *));
+    argv++;
+    argc--;
+
+    int i;
+    for(i = 0; i < req->r.shardgroups_replace.len; i++) {
+        ShardGroup *sg = RedisModule_Calloc(1, sizeof(ShardGroup));
+        req->r.shardgroups_replace.shardgroups[i] = sg;
+
+        long long num_slots, num_nodes;
+        if (RedisModule_StringToLongLong(argv[1], &num_slots) != REDISMODULE_OK ) {
+            RedisModule_ReplyWithError(ctx, "ERR invalid shard group message (num slots)");
+            goto fail;
+        }
+        if (RedisModule_StringToLongLong(argv[2], &num_nodes) != REDISMODULE_OK) {
+            RedisModule_ReplyWithError(ctx, "ERR invalid shard group message (num nodes)");
+            goto fail;
+        }
+
+        int num_argv_entries = (int) (3 + num_slots * 3 + num_nodes * 2);
+        if (num_argv_entries > argc) {
+            RedisModule_ReplyWithError(ctx, "ERR invalid shard group message (num entries)");
+            goto fail;
+        }
+
+        if (ShardGroupParse(ctx, argv, num_argv_entries, sg) != RR_OK) {
+            goto fail;
+        }
+
+        argv += num_argv_entries;
+        argc -= num_argv_entries;
+    }
+
+    return RR_OK;
+
+fail:
+    for (; i >= 0; i--) {
+        RedisModule_Free(req->r.shardgroups_replace.shardgroups[i]);
```

Comment:
I think we should also call ShardGroupFree() here. 

This issue aside, I think `ShardGroupFree()` creates a confusion because of its naming (It actually doesn't free ShardGroup itself).

Maybe we create these functions : 
```c
void ShardGroupInit(ShardGroup *sg) {
    memset(sg, 0, sizeof(ShardGroup));
}

void ShardGroupTerm(ShardGroup *sg)
{
    if (sg->conn) {
        ConnAsyncTerminate(sg->conn);
        sg->conn = NULL;
    }

    if (sg->slot_ranges) {
        RedisModule_Free(sg->slot_ranges);
        sg->slot_ranges = NULL;
    }

    if (sg->nodes) {
        RedisModule_Free(sg->nodes);
        sg->nodes = NULL;
    }
}

ShardGroup *ShardGroupCreate() {
    ShardGroup *sg = RedisModule_Alloc(sizeof(*sg));
    ShardGroupInit(sg);
    return sg;
}

void ShardGroupFree(ShardGroup *sg) {
    ShardGroupTerm(sg);
    RedisModule_Free(sg);
}
```

Now, we can pair `ShardGroupCreate()` with `ShardGroupFree()` and `ShardGroupInit()` with `ShardGroupTerm()`. We don't call manually RedisModule_Calloc() to create ShardGroup etc. All goes through these functions. We just make sure we pair these correctly. Does that make sense?

Reply:
yes

Reply:
added this.